auth/src/providers/AzureSubscriptionProviderBase.ts

```diff
@@ -0,0 +1,338 @@
+import type { AzureTenant } from '../contracts/AzureTenant';
+import { getConfiguredAuthProviderId, getConfiguredAzureEnv } from '../utils/configuredAzureEnv';
+import { dedupeSubscriptions } from '../utils/dedupeSubscriptions';
+import { getSessionFromVSCode } from '../utils/getSessionFromVSCode';
+import { isAuthenticationWwwAuthenticateRequest } from '../utils/isAuthenticationWwwAuthenticateRequest';
+import { isNotSignedInError, NotSignedInError } from '../utils/NotSignedInError';
+
+const EventDebounce = 5 * 1000; // 5 seconds minimum between `onRefreshSuggested` events
+const EventSilenceTime = 5 * 1000; // 5 seconds after sign-in to silence `onRefreshSuggested` events
+
+let armSubs: typeof import('@azure/arm-resources-subscriptions') | undefined;
+
+/**
+ * Base class for Azure subscription providers that use VS Code authentication.
+ * Handles actual communication with Azure via the Azure SDK, as well as
+ * controlling the firing of `onRefreshSuggested` events.
+ */
+export abstract class AzureSubscriptionProviderBase implements AzureSubscriptionProvider {
+    /**
+     * Constructs a new {@link AzureSubscriptionProviderBase}.
+     * @param logger (Optional) A logger to record information to
+     */
+    public constructor(protected readonly logger?: vscode.LogOutputChannel) { }
+
+    private lastRefreshSuggestedTime: number = 0;
+    private suppressRefreshSuggestedEvents: boolean = false;
+
+    /**
+     * @inheritdoc
+     */
+    public onRefreshSuggested(callback: () => any, thisArg?: any, disposables?: vscode.Disposable[]): vscode.Disposable { // eslint-disable-line @typescript-eslint/no-explicit-any -- Want to match the VSCode API
+        const disposable = vscode.authentication.onDidChangeSessions((evt) => {
+            if (evt.provider.id !== getConfiguredAuthProviderId()) {
+                // If it's not for the configured provider, ignore it
+                return;
+            }
+
+            if (this.suppressRefreshSuggestedEvents || Date.now() < this.lastRefreshSuggestedTime + EventDebounce) {
+                // Suppress and/or debounce events to avoid flooding
+                return;
+            }
+
+            this.logger?.debug('auth: Firing onRefreshSuggested event');
+
+            // Call the callback asynchronously to avoid potential issues
+            const timeout = setTimeout(() => {
+                clearTimeout(timeout);
+                this.lastRefreshSuggestedTime = Date.now();
+                // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+                void callback.bind(thisArg)();
+            }, 0);
+        });
+
+        disposables?.push(disposable);
+        return disposable;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public async signIn(tenant?: TenantIdAndAccount, options: SignInOptions = DefaultSignInOptions): Promise<boolean> {
+        this.silenceRefreshEvents();
```

Comment:
```suggestion
        if (!options.silent) {
            // Suppress without timeout, until sign in is done--it can take a while when done interactively
            this.suppressRefreshSuggestedEvents = true;
        } else {
            this.silenceRefreshEvents();
        }
```